packages/react/src/elements/components/Banner/BannerPresenter/index.js

```diff
@@ -0,0 +1,177 @@
+    /** The displayed message */
+    message: PropTypes.string,
+    /** Accessibility text for the dismiss button */
+    dismissButtonTitle: PropTypes.string,
+    /** Called when the banner is dismissed */
+    onDismiss: PropTypes.func,
+    /** The ID used for ARIA labeling */
+    labelId: PropTypes.string,
+    /** A render prop; Renders banner interactions */
+    children: PropTypes.func
+  };
+
+  /** @type {BannerPresenterState} */
+  state = {
+    isWrapping: false
+  };
+
+  /** @todo Add listener based on `children` prop */
+  componentWillMount() {
+    window.addEventListener("resize", this.handleResize);
+  }
+
+  componentDidMount() {
+    this.updateWrapping(this.triggerReady);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener("resize", this.handleResize);
+    window.cancelAnimationFrame(this.animationFrame);
+  }
+
+  /** @type {BannerPresenterProps} */
+  props;
+
+  /** @type {number} */
+  animationFrame;
+
+  /** @type {HTMLDivElement} */
+  notification;
+
+  /** @type {HTMLDivElement} */
+  interactionsWrapper;
+
+  /**
+   * @param {HTMLDivElement} notification
+   */
+  refNotification = notification => {
+    this.notification = notification;
+  };
+
+  /**
+   * @param {HTMLDivElement} interactionsWrapper
+   */
+  refInteractionsWrapper = interactionsWrapper => {
+    this.interactionsWrapper = interactionsWrapper;
+  };
+
+  handleResize = () => {
+    if (this.animationFrame !== undefined) return;
+
+    this.animationFrame = window.requestAnimationFrame(this.updateWrapping);
+  };
+
+  updateWrapping = () => {
+    /** @todo Refactor so that this function isn't called if the references don't exist */
+    const { notification, interactionsWrapper } = this;
+
+    if (!notification || !interactionsWrapper) return;
+
+    const isWrapping =
+      notification.offsetLeft === interactionsWrapper.offsetLeft;
+
+    if (isWrapping !== this.state.isWrapping) {
+      this.setState({ isWrapping });
+    }
+
+    delete this.animationFrame;
+  };
+
+  render() {
+    const {
+      type,
+      placement,
+      label,
+      message,
+      dismissButtonTitle,
+      onDismiss,
+      labelId,
+      children: renderActions
+    } = this.props;
+
+    const hasLabel = !!label;
+    const hasActions = !!renderActions;
+    const wrapperLabelledBy = hasLabel ? labelId : undefined;
+    const { isWrapping } = this.state;
+
+    return (
+      <Wrapper
+        type={type}
+        placement={placement}
+        hasActions={hasActions}
+        isWrapping={isWrapping}
+        labelledBy={wrapperLabelledBy}
+      >
+        <Icon type={type} />
+        <Content>
+          <Notification innerRef={this.refNotification}>
+            {hasLabel ? <Label id={labelId}>{label}</Label> : null}
+            <Message>{message}</Message>
+          </Notification>
+          {hasActions ? (
+            <InteractionsWrapper innerRef={this.refInteractionsWrapper}>
+              {renderActions({ isWrapping })}
```

Comment:
The `children` prop of the `Banner` component is a render prop. This allows consumers to render content within the banner using the `Banner`'s state.

```js
function CustomBanner() {
  return (
    <Banner>
      {({ isWrapping }) => <CustomButton isWrapping={isWrapping} />}
    </Banner>
  );
}
```

The current vision is too also provide various helper functions to create render props for common usecases.

```js
const renderConfirmation = Banner.createConfirmationRenderer({
  resolveText: "Accept changes",
  rejectText: "Decline changes",
  onResolve: () => console.log("User confirmed changed"),
  onReject: () => console.log("User declined changes")
});

function ChangesConfirmationBanner() {
  return <Banner>{this.renderConfirmation}</Banner>;
}
```